code/modules/power/pt_laser.dm

```diff
@@ -76,6 +82,52 @@
 
 	..()
 
+/obj/machinery/power/pt_laser/proc/_toggle_input_mechchomp()
+	src.charging = !src.charging
+
+/obj/machinery/power/pt_laser/proc/_set_input_mechchomp(var/datum/mechanicsMessage/inp)
+	if(!length(inp.signal)) return
+	var/newinput = text2num(inp.signal)
+	if (newinput != src.chargelevel && isnum_safe(newinput) && newinput > 0)
+		src.chargelevel = newinput
+		// Working backwards to update the ui based on the power we've set up.
+		if(chargelevel < 1 KILO WATT)
+			src.input_multi = 1 WATT
+		else if(chargelevel < 1 MEGA WATT)
+			src.input_multi = 1 KILO WATT
+		else if(chargelevel < 1 GIGA WATT)
+			src.input_multi = 1 MEGA WATT
+		else if(chargelevel < 1 TERA WATT)
+			src.input_multi = 1 GIGA WATT
+		else
+			src.input_multi = 1 TERA WATT
+		src.input_number = clamp((src.chargelevel/src.input_multi), 0, src.max_dial_value)
+
+/obj/machinery/power/pt_laser/proc/_toggle_output_mechchomp()
+	src.online = !src.online
+	if (!src.online && src.firing)
+		src.stop_firing()
+
+
+/obj/machinery/power/pt_laser/proc/_set_output_mechchomp(var/datum/mechanicsMessage/inp)
+	if(!length(inp.signal)) return
+	var/newoutput = text2num(inp.signal)
+	// We check against the absolute value of the current charge level, in case the PTL has been emagged.
+	if (newoutput != abs(src.output) && isnum_safe(newoutput) && newoutput > 0)
+		src.output = src.emagged ? -newoutput : newoutput
+		// Working backwards to update the ui based on the power we've set up.
+		if(newoutput >= 1 TERA WATT)
+			src.output_multi = 1 TERA WATT
+		else if(newoutput >= 1 GIGA WATT)
+			src.output_multi = 1 GIGA WATT
+		else
+			src.output_multi = 1 MEGA WATT
+		var abs_output_number = clamp((newoutput/src.output_multi), 0, src.max_dial_value)
+		src.output_number = src.emagged ? -abs_output_number : abs_output_number
+		src.update_output()
+		src.update_laser_power()
```

Comment:
Actually doesn't this still let you call `melt_blocking_objects` rapidly by flickering the value from `0` to `lots` repeatedly? It should probably have the same delayed startup behaviour that the toggle signal does.

Reply:
I appreciate the quick responses. Let me go through the whole file and fix this on the ui controls as well.

Reply:
I ran into some issues and ended up increasing the scope of the original changes. Two big problems were:
- Once I removed a lot of the redundant calls and pushed all the laser updates into the `process` function, I realized that the PTL was designed to only run off battery power (the `can_fire` proc checks to see `abs(src.output) <= src.charge`, and there are checks in the process loop as well). This made the previous fix where we reduced the battery capacity break things.
- If we defer the all laser changes to the process loop, the PTL gets a lot less responsive compared to now which felt bad. Even on my local, the delay was noticeable, and I imagine its only worse on live.

To fix the first issue, I refactored the `can_fire` proc and `process` to be able to check against the total available input power in addition to the charge (this will the PTL work with the reduced capacity).

To fix the second I added an adjustable cooldown to all PTL laser updates (currently `1 SECOND` but I can adjust that). It can toggle beam propagation instantly on the first click, but any clicks after will defer the updates to the next process loop update. This should stop from flickering the value.

This was my attempt to fix things based on my superficial knowledge of the codebase, and might be a little too amibitious. If you think I should defer all the changes to the process loop I can push a new revision.